tunnel/forwarder.go

```diff
@@ -0,0 +1,225 @@
+package tunnel
+
+import (
+	"fmt"
+	"github.com/google/uuid"
+	"github.com/hightouchio/passage/stats"
+	"github.com/pkg/errors"
+	"io"
+	"net"
+	"sync"
+	"time"
+)
+
+type TCPForwarder struct {
+	BindAddr string
+
+	// KeepaliveInterval is the interval between OS level TCP keepalive handshakes
+	KeepaliveInterval time.Duration
+
+	// GetUpstreamConn is a function thats job is to initiate a connection to the upstream service.
+	// It is called once for each incoming TunnelConnection.
+	// It should return a dedicated io.ReadWriteCloser for each incoming TunnelConnection.
+	GetUpstreamConn func(net.Conn) (io.ReadWriteCloser, error)
+
+	Lifecycle Lifecycle
+	Stats     stats.Stats
+
+	listener  *net.TCPListener
+	conns     map[string]net.Conn
+	close     chan struct{}
+	closeOnce sync.Once
+
+	sync.RWMutex
+}
+
+type TCPSession struct {
+	*net.TCPConn
+	id string
+
+	bytesSent, bytesReceived uint64
+}
+
+func (s *TCPSession) ID() string {
+	return s.id
+}
+
+func (f *TCPForwarder) Listen() error {
+	f.close = make(chan struct{})
+
+	// Open tunnel TCP listener
+	f.Lifecycle.BootEvent("listener_start", stats.Tags{"listen_addr": f.BindAddr})
+	listenTcpAddr, err := net.ResolveTCPAddr("tcp", f.BindAddr)
+	if err != nil {
+		return bootError{event: "listener_start", err: errors.Wrap(err, "resolve addr")}
+	}
+	listener, err := net.ListenTCP("tcp", listenTcpAddr)
+	if err != nil {
+		return bootError{event: "listener_start", err: err}
+	}
+	f.listener = listener
+	f.Lifecycle.Open()
+
+	// Wait for close signal
+	go func() {
+		<-f.close
+		f.listener.Close()
+		f.Lifecycle.Close()
+	}()
+
+	return nil
+}
+
+func (f *TCPForwarder) Serve() error {
+	for {
+		if f.listener == nil {
+			return fmt.Errorf("cannot serve without first starting listener")
+		}
+
+		// Accept incoming tunnel connections
+		conn, err := f.listener.AcceptTCP()
+		if err != nil {
+			return errors.Wrap(err, "accept tcp")
+		}
+
+		// Configure keepalive
+		conn.SetKeepAlive(true)
+		conn.SetKeepAlivePeriod(f.KeepaliveInterval)
+
+		// Pass connections off to tunnel connection handler.
+		go func() {
+			session := &TCPSession{
+				TCPConn: conn,
+				id:      uuid.New().String(),
+			}
+
+			f.handleSession(session)
+		}()
+	}
+}
+
+func (f *TCPForwarder) Close() error {
+	f.closeOnce.Do(func() { close(f.close) })
+	return nil
+}
+
+// handleSession takes a TCPSession (backed by a net.TCPConn), then initiates an upstream connection to our forwarding backend
+// and forwards packets between the two.
+func (f *TCPForwarder) handleSession(session *TCPSession) {
+	f.Lifecycle.SessionEvent(session.ID(), "open", stats.Tags{"remote_addr": session.RemoteAddr().String()})
+
+	defer func() {
+		session.Close()
+		// Record pipeline metrics to logs and statsd
+		f.Stats.Count("bytes_rcvd", int64(session.bytesReceived), nil, 1)
```

Comment:
is it worth sending these at the very end of the connection vs real-time as they are read/sent? I don't know how long these connections are open, but if it's ~minutes (rather than ~seconds), this won't help us in debugging a large, ongoing spike, for example